code/variant/variant.cpp

```diff
@@ -16,16 +16,16 @@ struct Particle {
   virtual ~Particle() = default;
 };
 
-struct Electron : public Particle {
-  virtual void print() const { std::cout << "E" << std::endl; }
+struct Electron : Particle {
+  void print() const final { std::cout << "E\n"; }
 };
 
-struct Proton : public Particle {
-  virtual void print() const { std::cout << "P" << std::endl; }
+struct Proton : Particle {
+  void print() const final { std::cout << "P\n"; }
 };
 
-struct Neutron : public Particle {
-  virtual void print() const { std::cout << "N" << std::endl; }
+struct Neutron : Particle {
+  void print() const final { std::cout << "N\n"; }
 };
```

Comment:
In general I wonder why you want to put a final here (class or method). If someone would like to write a `PrttyPrintedNeutron` with a different print method, why would you forbid it ?

Reply:
As a coding guideline I think it's good practice to default to `final` unless the class is explicitly designed to be inherited from. Inheriting from classes that are not designed to be inherited from rarely leads to happiness.

With that said I'm ok with using `override` instead of `final`, just not `virtual` because then the compiler will not emit an error in case you write a wrong signature or misspell the method name.

Reply:
> unless the class is explicitly designed to be inherited from

Agreed, but my conception of the language is that by default a class should be designed to be inherited from. Only if you're not in that default case you should protect future users of the class using final. Here there is nothing at all preventing inheritance, so default applies and I would avoid final personally.

Anyway, no big deal. I would still drop the virtual in all cases as it's useless and put the override as it's useful for error reporting indeed :-)

Reply:
> by default a class should be designed to be inherited from

I know we are digressing, but now I'm curious :smile:  you declare all destructors virtual unless you have a reason not to?

Reply:
yes, absolutely

Reply:
And I very seldom have a reason actually

Reply:
interesting :smile:

Reply:
I think print is a bad example to be a final method, something like the mass of the particle might be a better candidate :P

Reply:
alright alright, the least controversial improvement is `override` instead of `virtual` then